### PR TITLE
[FIX] Fix some Bugs on Grazing Range regarding audit's findings

### PR DESCRIPTION
**Pull request by abraham-alpaca**
## Description
This is a fix PR for fixing some issues rewarding andit's findings
There are 5 fixes in this PR
- `pendingReward()'s lastRewardBlock cursor`
- `public` to `external ` in `initialized()`
- `emergencyWithdraw()` updated total staked
- `emergencyRewardWithdraw()` amount validation
- `nonreentrancy guard on emergencyWithdraw and emergencyRewardWithdraw`

## Why?
- Some issues contain medium/crucial severity.

### CODE REVIEW

File: contracts/6/token/GrazingRange.sol [hunk 1]
```diff
@@ -276,19 +281,21 @@ contract GrazingRange is OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe  {
     }
 
     // @notice Withdraw without caring about rewards. EMERGENCY ONLY.
-    function emergencyWithdraw(uint256 _campaignID) external {
+    function emergencyWithdraw(uint256 _campaignID) external nonReentrant {
         CampaignInfo storage campaign = campaignInfo[_campaignID];
         UserInfo storage user = userInfo[_campaignID][msg.sender];
         campaign.stakingToken.safeTransfer(address(msg.sender), user.amount);
+        campaign.totalStaked -= user.amount;
```
Comment (spicysquid168) [inline contracts/6/token/GrazingRange.sol:288]: use openzeppelin sub instead

File: contracts/6/token/GrazingRange.sol [hunk 2]
```diff
@@ -276,19 +281,21 @@ contract GrazingRange is OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe  {
...
         user.rewardDebt = 0;
         emit EmergencyWithdraw(msg.sender, user.amount, _campaignID);
     }
 
     // @notice Withdraw reward. EMERGENCY ONLY.
-    function emergencyRewardWithdraw(uint256 _campaignID, uint256 _amount, address _beneficiary) external onlyOwner {
+    function emergencyRewardWithdraw(uint256 _campaignID, uint256 _amount, address _beneficiary) external onlyOwner nonReentrant {
         CampaignInfo storage campaign = campaignInfo[_campaignID];
-        require(_amount < campaign.rewardToken.balanceOf(address(this)), "GrazingRange::emergencyRewardWithdraw::not enough token");
+        uint256 currentStakingPendingReward = _pendingReward(_campaignID, campaign.totalStaked, 0);
```
Comment (spicysquid168) [inline contracts/6/token/GrazingRange.sol:297]: use address(0) not just pass 0

Comment (abraham-alpaca) [reply contracts/6/token/GrazingRange.sol:297]: this one is a rewardDebt parameter helper
so this internal function can be called regardless the user
`_pendingReward(uint256 _campaignID, uint256 _amount, uint256 _rewardDebt)`

File: contracts/6/token/GrazingRange.sol (hunk 1 above)
Comment (abraham-alpaca) [reply contracts/6/token/GrazingRange.sol:288]: okay

File: contracts/6/token/GrazingRange.sol (hunk 2 above)
Comment (spicysquid168) [reply contracts/6/token/GrazingRange.sol:297]: Oh I see. Sorry. I was thinking it is `pending()`

**Review by spicysquid168 — APPROVED** [review]

**Review by jakeal-papaca — APPROVED** [review]

**Pull request merged**